### ISSUE TITLE
Dev server CSP items are missing when requested with Accept: text/html using Fastboot on master

### ISSUE THREAD

**Issue by reidab** [rwjblue/ember-cli-content-security-policy#133]
Since it looks like a beta of 2.x is coming soon, I was testing out the `master` branch (https://github.com/rwjblue/ember-cli-content-security-policy/commit/5dc741e3e7739899c5ee97eea5e86450becb1ede) to see if it would solve something similar to #127 that I was running into. It did, but I'm now running into some bizarre behavior around the `localhost` CSP injection for the dev server with Fastboot.

When requested with an `Accept` header specifying `*/*` (the default from `curl`), it looks like the raw `index.html` output is returned along with the expected CSP headers. When requesting with an `Accept` header that includes `text/html` (like a browser would use) the Fastboot-rendered body is returned but the `content-security-policy` header no longer includes `localhost:4200 0.0.0.0:4200` and other directives that were injected for local dev server usage. Strangely, the `x-content-security-policy` header still includes these values.

The calls to `setHeader` in ember-cli-content-security-policy certainly seem to be called with the appropriate values, so maybe something's rewriting the headers further downstream? I'd say it was failing to detect `liveReload`, but that wouldn't explain `x-content-security-policy` sneaking through unscathed…

Anyhow, I'm not sure if this issue is with ember-cli-content-security-policy or ember-cli-fastboot, but it was baffling enough that I wanted to write it up. 😁 

## Expected

Both `content-security-policy` and `x-content-security-policy` to be the same CSP string including `script-src` and `connect-src` entries for localhost.

## Actual

`x-content-security-policy` is delivered with the localhost entries included

```
x-content-security-policy: default-src 'none'; script-src 'self' localhost:4200
0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200
ws://0.0.0.0:4200; img-src 'self'; style-src 'self'; media-src 'self';
```

`content-security-policy` is delivered without the localhost entries

```
content-security-policy: default-src 'none'; script-src 'self'; font-src 'self'; 
connect-src 'self'; img-src 'self'; style-src 'self'; media-src 'self';
```

## Repro / Investigation

```
$ ember new csp-test
installing app
Ember CLI v3.17.0
[…]

$ cd csp-test
$ ember install ember-cli-content-security-policy@1.1.1
$ ember s
```

At this point, I'm running 1.1.1 without Fastboot. CSP headers are returned as expected regardless of the `Accept` header.

<details>
  <summary><code>$ curl http://localhost:4200</code> (click to expand)</summary>

```  
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4200 (#0)
> GET / HTTP/1.1
> Host: localhost:4200
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< X-Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< Cache-Control: private, max-age=0, must-revalidate
< Last-Modified: Thu, 19 Mar 2020 04:13:04 GMT
< Content-Length: 1613
< Content-Type: text/html; charset=utf-8
< Vary: Accept-Encoding
< Date: Thu, 19 Mar 2020 04:21:38 GMT
< Connection: keep-alive
<
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>CspTest</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="csp-test/config/environment" content="%7B%22modulePrefix%22%3A%22csp-test%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22csp-test%22%2C%22version%22%3A%220.0.0%2Be81a7384%22%7D%2C%22contentSecurityPolicyHeader%22%3A%22Content-Security-Policy-Report-Only%22%2C%22contentSecurityPolicy%22%3A%7B%22default-src%22%3A%5B%22'none'%22%5D%2C%22script-src%22%3A%5B%22'self'%22%5D%2C%22font-src%22%3A%5B%22'self'%22%5D%2C%22connect-src%22%3A%5B%22'self'%22%5D%2C%22img-src%22%3A%5B%22'self'%22%5D%2C%22style-src%22%3A%5B%22'self'%22%5D%2C%22media-src%22%3A%5B%22'self'%22%5D%7D%2C%22exportApplicationGlobal%22%3Atrue%7D" />
<script src="/ember-cli-live-reload.js" type="text/javascript"></script>

    <link integrity="" rel="stylesheet" href="/assets/vendor.css">
    <link integrity="" rel="stylesheet" href="/assets/csp-test.css">


  </head>
  <body>


    <script src="/assets/vendor.js"></script>
    <script src="/assets/csp-test.js"></script>


  </body>
</html>
* Connection #0 to host localhost left intact
* Closing connection 0
```
</details>

<details>
  <summary><code>$ curl 'http://localhost:4200/' -H 'Accept: text/html' -v</code> (click to expand)</summary>

```  
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4200 (#0)
> GET / HTTP/1.1
> Host: localhost:4200
> User-Agent: curl/7.64.1
> Accept: text/html
>
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< X-Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< Cache-Control: private, max-age=0, must-revalidate
< Last-Modified: Thu, 19 Mar 2020 04:13:04 GMT
< Content-Length: 1613
< Content-Type: text/html; charset=utf-8
< Vary: Accept-Encoding
< Date: Thu, 19 Mar 2020 04:23:50 GMT
< Connection: keep-alive
<
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>CspTest</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="csp-test/config/environment" content="%7B%22modulePrefix%22%3A%22csp-test%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22csp-test%22%2C%22version%22%3A%220.0.0%2Be81a7384%22%7D%2C%22contentSecurityPolicyHeader%22%3A%22Content-Security-Policy-Report-Only%22%2C%22contentSecurityPolicy%22%3A%7B%22default-src%22%3A%5B%22'none'%22%5D%2C%22script-src%22%3A%5B%22'self'%22%5D%2C%22font-src%22%3A%5B%22'self'%22%5D%2C%22connect-src%22%3A%5B%22'self'%22%5D%2C%22img-src%22%3A%5B%22'self'%22%5D%2C%22style-src%22%3A%5B%22'self'%22%5D%2C%22media-src%22%3A%5B%22'self'%22%5D%7D%2C%22exportApplicationGlobal%22%3Atrue%7D" />
<script src="/ember-cli-live-reload.js" type="text/javascript"></script>

    <link integrity="" rel="stylesheet" href="/assets/vendor.css">
    <link integrity="" rel="stylesheet" href="/assets/csp-test.css">


  </head>
  <body>


    <script src="/assets/vendor.js"></script>
    <script src="/assets/csp-test.js"></script>


  </body>
</html>
* Connection #0 to host localhost left intact
* Closing connection 0
```
</details>

---

```
$ ember install ember-cli-fastboot
$ ember s
```

After adding Fastboot, we start to see a difference between the responses when requested with `Accept: */*` and `Accept: text/html`. With `text/html` we get the rendered Fastboot body, while with `*/*` we seem to get the raw index.html with comment placeholders for the Fastboot content. At this point, the CSP headers are still rendering as expected.

<details>
  <summary><code>$ curl 'http://localhost:4200/ -v</code> (click to expand)</summary>

```  
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4200 (#0)
> GET / HTTP/1.1
> Host: localhost:4200
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< X-Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< Cache-Control: private, max-age=0, must-revalidate
< Last-Modified: Thu, 19 Mar 2020 04:26:38 GMT
< Content-Length: 1711
< Content-Type: text/html; charset=utf-8
< Vary: Accept-Encoding
< Date: Thu, 19 Mar 2020 04:26:49 GMT
< Connection: keep-alive
<
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>CspTest</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="csp-test/config/environment" content="%7B%22modulePrefix%22%3A%22csp-test%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22csp-test%22%2C%22version%22%3A%220.0.0%2Be81a7384%22%7D%2C%22contentSecurityPolicyHeader%22%3A%22Content-Security-Policy-Report-Only%22%2C%22contentSecurityPolicy%22%3A%7B%22default-src%22%3A%5B%22'none'%22%5D%2C%22script-src%22%3A%5B%22'self'%22%5D%2C%22font-src%22%3A%5B%22'self'%22%5D%2C%22connect-src%22%3A%5B%22'self'%22%5D%2C%22img-src%22%3A%5B%22'self'%22%5D%2C%22style-src%22%3A%5B%22'self'%22%5D%2C%22media-src%22%3A%5B%22'self'%22%5D%7D%2C%22exportApplicationGlobal%22%3Atrue%7D" />
<script src="/ember-cli-live-reload.js" type="text/javascript"></script>
<!-- EMBER_CLI_FASTBOOT_TITLE --><!-- EMBER_CLI_FASTBOOT_HEAD -->

    <link integrity="" rel="stylesheet" href="/assets/vendor.css">
    <link integrity="" rel="stylesheet" href="/assets/csp-test.css">


  </head>
  <body>
    <!-- EMBER_CLI_FASTBOOT_BODY -->

    <script src="/assets/vendor.js"></script>
    <script src="/assets/csp-test.js"></script>


  </body>
</html>
* Connection #0 to host localhost left intact
* Closing connection 0
```
</details>

<details>
  <summary><code>$ curl 'http://localhost:4200/' -H 'Accept: text/html' -v</code> (click to expand)</summary>

```  
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4200 (#0)
> GET / HTTP/1.1
> Host: localhost:4200
> User-Agent: curl/7.64.1
> Accept: text/html
>
< HTTP/1.1 200 OK
< x-powered-by: Express
< content-security-policy-report-only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< x-content-security-policy-report-only: default-src 'none'; script-src 'self' 'nonce-abcdefg' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< cache-control: private, max-age=0, must-revalidate
< Content-Type: text/html; charset=utf-8
< Content-Length: 3678
< ETag: W/"e5e-cxMlOlCJ76g+S84d3dFGKGrUrTQ"
< Vary: Accept-Encoding
< Date: Thu, 19 Mar 2020 04:27:51 GMT
< Connection: keep-alive
<
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>CspTest</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="csp-test/config/environment" content="%7B%22modulePrefix%22%3A%22csp-test%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22csp-test%22%2C%22version%22%3A%220.0.0%2Be81a7384%22%7D%2C%22contentSecurityPolicyHeader%22%3A%22Content-Security-Policy-Report-Only%22%2C%22contentSecurityPolicy%22%3A%7B%22default-src%22%3A%5B%22'none'%22%5D%2C%22script-src%22%3A%5B%22'self'%22%5D%2C%22font-src%22%3A%5B%22'self'%22%5D%2C%22connect-src%22%3A%5B%22'self'%22%5D%2C%22img-src%22%3A%5B%22'self'%22%5D%2C%22style-src%22%3A%5B%22'self'%22%5D%2C%22media-src%22%3A%5B%22'self'%22%5D%7D%2C%22exportApplicationGlobal%22%3Atrue%7D" />
<script src="/ember-cli-live-reload.js" type="text/javascript"></script>
<!-- EMBER_CLI_FASTBOOT_TITLE -->

    <link integrity="" rel="stylesheet" href="/assets/vendor.css">
    <link integrity="" rel="stylesheet" href="/assets/csp-test.css">


  </head>
  <body>
    <script type="x/boundary" id="fastboot-body-start"></script><div id="ember180" class="ember-view"><div id="ember-welcome-page-id-selector" data-ember-version="3.17.0">
  <div class="columns">
    <div class="tomster">
      <img src="ember-welcome-page/images/construction.png" alt="Under construction">
    </div>
    <div class="welcome">
      <h2 id="title">Congratulations, you made it!</h2>

      <p>You’ve officially spun up your very first Ember app :-)</p>
      <p>You’ve got one more decision to make: what do you want to do next? We’d suggest one of the following to help you get going:</p>
      <ol>
        <li><a href="https://guides.emberjs.com/v3.17.0/getting-started/quick-start/">Quick Start</a> - a quick introduction to how Ember works. Learn about defining your first route, writing a UI component and deploying your application.</li>
        <li><a href="https://guides.emberjs.com/v3.17.0/tutorial/ember-cli/">Ember Guides</a> - this is our more thorough, hands-on intro to Ember. Your crash course in Ember philosophy, background and some in-depth discussion of how things work (and why they work the way they do).</li>
      </ol>
      <p>If you run into problems, you can check <a href="http://stackoverflow.com/questions/tagged/ember.js">Stack Overflow</a> or <a href="http://discuss.emberjs.com/">our forums</a>  for ideas and answers—someone’s probably been through the same thing and already posted an answer.  If not, you can post your <strong>own</strong> question. People love to help new Ember developers get started, and our <a href="https://emberjs.com/community/">Ember Community</a> is incredibly supportive.</p>
    </div>
  </div>
  <p class="postscript">To remove this welcome message, remove the
    <code>
        &lt;WelcomePage /&gt;
    </code>
    component from your
    <code>
        app/templates/application.hbs
    </code>
     file.<br>You'll see this page update soon after!
  </p>
</div>
</div>

<!----><script type="x/boundary" id="fastboot-body-end"></script>

    <script src="/assets/vendor.js"></script>
    <script src="/assets/csp-test.js"></script>


  </body>
</html>
* Connection #0 to host localhost left intact
* Closing connection 0
```
</details>

---

```
$ ember install ember-cli-content-security-policy@rwjblue/ember-cli-content-security-policy#master
```

After switching to `master`, we start getting different CSP header responses between the two `Accept` headers in addition to the difference in page content.

<details>
  <summary><code>$ curl 'http://localhost:4200/ -v</code> (click to expand) — <strong>includes localhost:4200 0.0.0.0:4200 as expected</strong></summary>

```  
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4200 (#0)
> GET / HTTP/1.1
> Host: localhost:4200
> User-Agent: curl/7.64.1
> Accept: */*
>
< HTTP/1.1 200 OK
< X-Powered-By: Express
< Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< X-Content-Security-Policy-Report-Only: default-src 'none'; script-src 'self' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< Cache-Control: private, max-age=0, must-revalidate
< Last-Modified: Thu, 19 Mar 2020 04:34:07 GMT
< Content-Length: 1585
< Content-Type: text/html; charset=utf-8
< Vary: Accept-Encoding
< Date: Thu, 19 Mar 2020 04:34:10 GMT
< Connection: keep-alive
<
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>CspTest</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="csp-test/config/environment" content="%7B%22modulePrefix%22%3A%22csp-test%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22csp-test%22%2C%22version%22%3A%220.0.0%2Be81a7384%22%7D%2C%22ember-cli-content-security-policy%22%3A%7B%22policy%22%3A%22default-src%20'none'%3B%20script-src%20'self'%3B%20font-src%20'self'%3B%20connect-src%20'self'%3B%20img-src%20'self'%3B%20style-src%20'self'%3B%20media-src%20'self'%3B%22%2C%22reportOnly%22%3Atrue%7D%2C%22exportApplicationGlobal%22%3Atrue%7D" />
<script src="/ember-cli-live-reload.js" type="text/javascript"></script>
<!-- EMBER_CLI_FASTBOOT_TITLE --><!-- EMBER_CLI_FASTBOOT_HEAD -->

    <link integrity="" rel="stylesheet" href="/assets/vendor.css">
    <link integrity="" rel="stylesheet" href="/assets/csp-test.css">


  </head>
  <body>
    <!-- EMBER_CLI_FASTBOOT_BODY -->

    <script src="/assets/vendor.js"></script>
    <script src="/assets/csp-test.js"></script>


  </body>
</html>
* Connection #0 to host localhost left intact
* Closing connection 0
```
</details>

<details>
  <summary><code>$ curl 'http://localhost:4200/' -H 'Accept: text/html' -v</code> (click to expand) — <strong>localhost:4200 0.0.0.0:4200 are missing from content-security-policy, but included in x-content-security-policy</strong></summary>

```  
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 4200 (#0)
> GET / HTTP/1.1
> Host: localhost:4200
> User-Agent: curl/7.64.1
> Accept: text/html
>
< HTTP/1.1 200 OK
< x-powered-by: Express
< content-security-policy-report-only: default-src 'none'; script-src 'self'; font-src 'self'; connect-src 'self'; img-src 'self'; style-src 'self'; media-src 'self';
< x-content-security-policy-report-only: default-src 'none'; script-src 'self' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
< cache-control: private, max-age=0, must-revalidate
< Content-Type: text/html; charset=utf-8
< Content-Length: 3552
< ETag: W/"de0-bdw9X5P68MGVLGegTCEETlvoTtw"
< Vary: Accept-Encoding
< Date: Thu, 19 Mar 2020 04:36:15 GMT
< Connection: keep-alive
<
<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>CspTest</title>
    <meta name="description" content="">
    <meta name="viewport" content="width=device-width, initial-scale=1">


<meta name="csp-test/config/environment" content="%7B%22modulePrefix%22%3A%22csp-test%22%2C%22environment%22%3A%22development%22%2C%22rootURL%22%3A%22%2F%22%2C%22locationType%22%3A%22auto%22%2C%22EmberENV%22%3A%7B%22FEATURES%22%3A%7B%7D%2C%22EXTEND_PROTOTYPES%22%3A%7B%22Date%22%3Afalse%7D%2C%22_APPLICATION_TEMPLATE_WRAPPER%22%3Afalse%2C%22_DEFAULT_ASYNC_OBSERVERS%22%3Atrue%2C%22_JQUERY_INTEGRATION%22%3Afalse%2C%22_TEMPLATE_ONLY_GLIMMER_COMPONENTS%22%3Atrue%7D%2C%22APP%22%3A%7B%22name%22%3A%22csp-test%22%2C%22version%22%3A%220.0.0%2Be81a7384%22%7D%2C%22ember-cli-content-security-policy%22%3A%7B%22policy%22%3A%22default-src%20'none'%3B%20script-src%20'self'%3B%20font-src%20'self'%3B%20connect-src%20'self'%3B%20img-src%20'self'%3B%20style-src%20'self'%3B%20media-src%20'self'%3B%22%2C%22reportOnly%22%3Atrue%7D%2C%22exportApplicationGlobal%22%3Atrue%7D" />
<script src="/ember-cli-live-reload.js" type="text/javascript"></script>
<!-- EMBER_CLI_FASTBOOT_TITLE -->

    <link integrity="" rel="stylesheet" href="/assets/vendor.css">
    <link integrity="" rel="stylesheet" href="/assets/csp-test.css">


  </head>
  <body>
    <script type="x/boundary" id="fastboot-body-start"></script><div id="ember180" class="ember-view"><div id="ember-welcome-page-id-selector" data-ember-version="3.17.0">
  <div class="columns">
    <div class="tomster">
      <img src="ember-welcome-page/images/construction.png" alt="Under construction">
    </div>
    <div class="welcome">
      <h2 id="title">Congratulations, you made it!</h2>

      <p>You’ve officially spun up your very first Ember app :-)</p>
      <p>You’ve got one more decision to make: what do you want to do next? We’d suggest one of the following to help you get going:</p>
      <ol>
        <li><a href="https://guides.emberjs.com/v3.17.0/getting-started/quick-start/">Quick Start</a> - a quick introduction to how Ember works. Learn about defining your first route, writing a UI component and deploying your application.</li>
        <li><a href="https://guides.emberjs.com/v3.17.0/tutorial/ember-cli/">Ember Guides</a> - this is our more thorough, hands-on intro to Ember. Your crash course in Ember philosophy, background and some in-depth discussion of how things work (and why they work the way they do).</li>
      </ol>
      <p>If you run into problems, you can check <a href="http://stackoverflow.com/questions/tagged/ember.js">Stack Overflow</a> or <a href="http://discuss.emberjs.com/">our forums</a>  for ideas and answers—someone’s probably been through the same thing and already posted an answer.  If not, you can post your <strong>own</strong> question. People love to help new Ember developers get started, and our <a href="https://emberjs.com/community/">Ember Community</a> is incredibly supportive.</p>
    </div>
  </div>
  <p class="postscript">To remove this welcome message, remove the
    <code>
        &lt;WelcomePage /&gt;
    </code>
    component from your
    <code>
        app/templates/application.hbs
    </code>
     file.<br>You'll see this page update soon after!
  </p>
</div>
</div>

<!----><script type="x/boundary" id="fastboot-body-end"></script>

    <script src="/assets/vendor.js"></script>
    <script src="/assets/csp-test.js"></script>


  </body>
</html>
* Connection #0 to host localhost left intact
* Closing connection 0
```
</details>

---

If I inject logging into ember-cli-content-security-policy's `index.js` immediately after the calls to `setHeader`, I can see in the server logs that both calls are made with the same correct CSP string.

```
setHeader Content-Security-Policy-Report-Only default-src 'none'; script-src 'self' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
setHeader X-Content-Security-Policy-Report-Only default-src 'none'; script-src 'self' localhost:4200 0.0.0.0:4200; font-src 'self'; connect-src 'self' ws://localhost:4200 ws://0.0.0.0:4200 http://localhost:4200; img-src 'self'; style-src 'self'; media-src 'self'; report-uri http://localhost:4200/csp-report;
```

**Comment by jelhan** [comment]
I haven't reproduced myself but my first guess would be a conflict between FastBoot and the middleware used in Ember CLI's development server. Both are setting the `Content-Security-Policy` header:
https://github.com/rwjblue/ember-cli-content-security-policy/blob/5dc741e3e7739899c5ee97eea5e86450becb1ede/index.js#L191-L199
https://github.com/rwjblue/ember-cli-content-security-policy/blob/5dc741e3e7739899c5ee97eea5e86450becb1ede/fastboot/instance-initializers/content-security-policy.js#L28-L29

If used together they overrule each other. It seems like the FastBoot initalizer wins.

It gets more complex as there seems to be two other issues:

1. FastBoot initializer only sets `Content-Security-Policy` header but not `X-Content-Security-Policy` header: https://github.com/rwjblue/ember-cli-content-security-policy/blob/5dc741e3e7739899c5ee97eea5e86450becb1ede/fastboot/instance-initializers/content-security-policy.js#L28-L29 This could explain the inconsistency between both.
2. The live reload support is injected into the config when the development server middleware runs first time: https://github.com/rwjblue/ember-cli-content-security-policy/blob/5dc741e3e7739899c5ee97eea5e86450becb1ede/index.js#L144-L159 But the config for FastBoot is already injected in `config` hook which runs before: https://github.com/rwjblue/ember-cli-content-security-policy/blob/5dc741e3e7739899c5ee97eea5e86450becb1ede/index.js#L115-L123 Therefore the CSP used by FastBoot misses the livereload support.

**Comment by jelhan** [comment]
Done the reproduction. Thanks a lot for the detailed steps!

The difference between `Accept: */*` and `Accept: text/html` header is caused by FastBoot. It seems like FastBoot does not run if the Accept header is `Accept: */*`. At least I don't see the rendered welcome page if called with `Accept: */*`. Not sure if this is intended or a bug.

FastBoot instance initializer is executed after development server middleware. It's documentation is not that good. But it states that [the header API implements parts of Fetch API's Header class](https://ember-fastboot.com/docs/user-guide#response-headers). The `Header.set` method, which we use, replaces existing headers. So that supports my assumption above.

To be honest I'm not sure if that's what we want to have. I'm tending towards not letting the FastBoot initalizier override an existing CSP header. We could assume that such a header is set by intention and overriding it might cause security issues.

**Comment by rwjblue** [comment]
Ya, agree

**Comment by jelhan** [comment]
@reidab Should be fixed by #136. It would be awesome if you could confirm from your side if you have the time to do so.
